### ISSUE TITLE
Flaky Test: Kestrel stops sending responses

### ISSUE THREAD

**Issue by mikeharder** [aspnet/KestrelHttpServer#2596]
# Failure
```
  Failed   ThrowsWhenUpgradingNonUpgradableRequest
   Assert.Equal() Failure
              (pos 0)
  Expected: HTTP/1.1 200 OK
  Actual:   
              (pos 0)
  Stack Trace:
     at Microsoft.AspNetCore.Testing.TestConnection.<Receive>d__23.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.UpgradeTests.<ThrowsWhenUpgradingNonUpgradableRequest>d__6.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Failed   TestDefaultHeaders
   Assert.Equal() Failure
              (pos 0)
  Expected: HTTP/1.1 200 OK\r\nDate: Sun, 20 May 2018 2úúú
  Actual:   
              (pos 0)
  Stack Trace:
     at Microsoft.AspNetCore.Testing.TestConnection.<Receive>d__23.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at Microsoft.AspNetCore.Testing.TestConnection.<ReceiveForcedEnd>d__25.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.DefaultHeaderTests.<TestDefaultHeaders>d__0.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

http://aspnetci/viewLog.html?buildId=471850&buildTypeId=Lite_KestrelStressTest&tab=buildResultsDiv

# Notes
The build failed due to a test hang.  Typically, when a test run is hung, all tests prior to the hang will have passed.  However, in this case it appears Kestrel got into a bad state where it stopped sending responses, which caused many tests to fail until one test finally hung.  Search the build log for `RUNDOTNET : error`.

**Comment by halter73** [comment]
Yikes! Here's the corresponding server-side log for the first failing test:

```
[7.804s] [Microsoft.AspNetCore.Server.Kestrel] [Warning] Connection processing ended abnormally.
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.SpanHelpers.IndexOf(Byte& searchSpace, Byte value, Int32 length)
   at System.MemoryExtensions.IndexOf[T](ReadOnlySpan`1 span, T value)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.ParseRequestLine(TRequestHandler handler, ReadOnlySequence`1& buffer, SequencePosition& consumed, SequencePosition& examined)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser`1.Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.IHttpParser<TRequestHandler>.ParseRequestLine(TRequestHandler handler, ReadOnlySequence`1& buffer, SequencePosition& consumed, SequencePosition& examined)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TakeStartLine(ReadOnlySequence`1 buffer, SequencePosition& consumed, SequencePosition& examined)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.ParseRequest(ReadOnlySequence`1 buffer, SequencePosition& consumed, SequencePosition& examined)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1Connection.TryParseRequest(ReadResult result, Boolean& endConnection)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.<ProcessRequests>d__186`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

The fact that all the subsequent functional tests failed despite being complete new server instances, I suspect heap corruption 😨

**Comment by benaadams** [comment]
Weird, can't even see where it would throw an exception in [`System.SpanHelpers.IndexOf(Byte& searchSpace, Byte value, Int32 length)`](https://github.com/dotnet/coreclr/blob/821317f109635cc591c976e6921429fc7f673b68/src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs#L106-L214)

**Comment by halter73** [comment]
What's crazy is that a bunch of previous tests passed, and after the first test fails with the exception, every test fails with this exact same exception. I'm wondering if the JITed SpanHelpers.IndexOf method got corrupted somehow.

**Comment by Tratcher** [comment]
Similar symptoms to https://github.com/aspnet/KestrelHttpServer/issues/2635

**Comment by halter73** [comment]
@Tratcher The difference here is the IndexOutOfRangeException in the logs the other #2635 related failures haven't included these logs when I looked. The only time I've seen this logged is with the test run linked to by the original issue.

**Comment by ryanbrandenburg** [comment]
This looks like the same thing as [this](http://aspnetci/viewLog.html?buildId=507772&buildTypeId=Lite_KestrelStressTest).

**Comment by halter73** [comment]
@ryanbrandenburg Thanks for the heads up. This actually looks like a new issue. #2732

**Comment by Eilon** [comment]
@halter73 - can you work on resolving this flakiness?

**Comment by halter73** [comment]
@Eilon This issue's tricky. The original issue was caused by an `IndexOutOfRangeException` being thrown on the server for every test case in a single test run. We've never seen that again. The only reasonable explanation for this is some sort of memory corruption. With only one repro 2 months ago, I don't think this is worth investigating further.

Some similar looking issues occur only in our libuv functional tests on macOS and rarely on linux. #2635 details what I've discovered so far about these failures. Basically, our logs prove libuv is reporting client FINs even though we know we haven't closed the managed socket yet. It's very odd, and I would like to investigate this more, but it could be days or weeks more work to figure anything out. I think it's more important to keep making progress on HTTP/2 for now.

The most recent test failure linked to by @ryanbrandenburg was unrelated, and I fixed it with #2732 which was merged a few minutes ago.
